src/conversations/Conversations.ts

```diff
@@ -97,56 +97,115 @@ export default class Conversations {
+      const contentTopic = env.contentTopic
+      if (!contentTopic) {
+        return null
+      }
+
+      if (contentTopic === introTopic) {
+        const msg = await this.client.decodeEnvelope(env)
+        if (!messageHasHeaders(msg)) {
+          return null
+        }
 
-    // If we detect a new intro topic, update the stream's direct message topics to include the new topic
-    const contentTopicUpdater = (msg: MessageV1): string[] | undefined => {
-      if (msg.contentTopic !== introTopic || !messageHasHeaders(msg)) {
-        return undefined
+        return msg
       }
-      const peerAddress = this.getPeerAddress(msg)
-      if (
-        dmAddresses.includes(peerAddress) ||
-        peerAddress === this.client.address
-      ) {
-        // No need to update if we're already subscribed
-        return undefined
+
+      // Decode as an invite and return the envelope
+      // This gives the contentTopicUpdater everything it needs to add to the topic list
+      if (contentTopic === inviteTopic) {
+        const sealed = await SealedInvitation.fromEnvelope(env)
+        const unsealed = await sealed.v1.getInvitation(this.client.keys)
+        return ConversationV2.create(this.client, unsealed, sealed.v1.header)
+      }
+
+      const convo = convoMap.get(contentTopic)
+
+      // Decode as a V1 message if the topic matches a V1 convo
+      if (convo instanceof ConversationV1) {
+        return this.client.decodeEnvelope(env)
       }
-      dmAddresses.push(peerAddress)
-      return this.buildTopicsForAllMessages(dmAddresses, introTopic)
+
+      // Decode as a V2 message if the topic matches a V2 convo
+      if (convo instanceof ConversationV2) {
+        return convo.decodeMessage(env)
+      }
+
+      console.log('Unknown topic')
+
+      throw new Error('Unknown topic')
     }
 
-    // Filter intro topics if already streaming direct messages for that address to avoid duplicates
-    const filter = (msg: MessageV1): boolean => {
-      if (
-        msg.contentTopic === introTopic &&
-        messageHasHeaders(msg) &&
-        dmAddresses.includes(this.getPeerAddress(msg))
-      ) {
-        return false
+    const contentTopicUpdater = (msg: Conversation | Message | null) => {
+      // If we have a V1 message from the introTopic, store the conversation in our mapping
+      if (msg instanceof MessageV1 && msg.contentTopic === introTopic) {
+        const convo = new ConversationV1(
+          this.client,
+          msg.recipientAddress === this.client.address
+            ? (msg.senderAddress as string)
+            : (msg.recipientAddress as string),
+          msg.sent
+        )
+
+        if (topics.has(convo.topic)) {
+          return undefined
+        }
+
+        convoMap.set(convo.topic, convo)
+        topics.add(convo.topic)
+
+        return Array.from(topics.values())
       }
-      return true
+
+      if (msg instanceof ConversationV2) {
+        if (topics.has(msg.topic)) {
```

Comment:
This is just a nitpick but 175-182 is the same as 164-171, is it worth combining?

Reply:
I refactored to simplify